src/main.rs

```diff
@@ -119,140 +130,150 @@ fn process_buffer(buf: &[u8]) -> (HashMap<String, WeatherData>, u32) {
-    }
+    });
 
-    let (station_temperatures, lines_count) = process_buffer(&buf[start_index..end_index]);
-
-    let mut total_lines = total_lines.lock().unwrap();
-    *total_lines += lines_count;
-    drop(total_lines);
+    return (station_temperatures, lines_count);
+}
 
-    return station_temperatures;
+fn process_thread(buf: &[u8], extra_buffer_size: usize) -> (StationTemperatures, u32) {
+    let start_index = buf
+        .iter()
+        .position(|&b| b == b'\n')
+        .map(|i| i + 1)
+        .unwrap_or(0);
+
+    let buf_default_pos = buf.len() - extra_buffer_size;
+    let end_index = buf[buf_default_pos..]
+        .iter()
+        .position(|&b| b == b'\n')
+        .map(|i| i + buf_default_pos + 1)
+        .unwrap_or(buf_default_pos);
+
+    process_buffer(&buf[start_index..end_index])
 }
 
+const TOTAL_LINES: usize = 1_000_000_000;
+const AVG_ROW_SIZE: usize = 14;
+const THREAD_COUNT: usize = 250;
+const BUFFER_SIZE: usize = 2_000_000;
+const STAGE_COUNT: usize = (TOTAL_LINES * AVG_ROW_SIZE).div_ceil(THREAD_COUNT * BUFFER_SIZE);
+const SINGLE_ROW_SIZE: usize = 64;
+
 fn main() {
     let start_time = time::Instant::now();
 
+    println!("buffer size: {:?}", BUFFER_SIZE);
+
     // let cores: usize = std::thread::available_parallelism().unwrap().into();
     // println!("{}", cores);
 
-    // let file_path = "weather_stations.csv";
+    //let file_path = "weather_stations.csv";
     let file_path = "measurements.txt";
 
-    let stage_count = 30;
-    let max_threads = 250;
-    let buffer_size = 2000000;
-    let single_row_size = 100;
-
-    let mut station_temperatures: HashMap<String, WeatherData> = HashMap::new();
+    let mut station_temperatures: StationTemperatures = AHashMap::with_capacity(500);
 
     // Process first line
     let mut file = File::open(file_path).expect("Unable to open file");
-    let mut buf = vec![0; single_row_size];
+    let mut buf = [0; KEY_SIZE + 5];
     file.seek(SeekFrom::Start(0)).unwrap();
     file.read(&mut buf).unwrap();
     let first_line = str::from_utf8(&buf)
         .unwrap()
         .split('\n')
         .collect::<Vec<&str>>()[0];
-    let first_line_data = process_weather_line(first_line).unwrap();
-    station_temperatures.insert(first_line_data.0.to_string(), first_line_data.1);
+    let (key, value) = process_weather_line(first_line);
+    station_temperatures.insert(key, value);
 
-    let total_lines = Arc::new(Mutex::new(1));
-    let station_temperatures_list: Arc<Mutex<Vec<HashMap<String, WeatherData>>>> =
-        Arc::new(Mutex::new(Vec::new()));
+    let total_lines = Arc::new(AtomicU32::new(1));
+    let station_temperatures_list: Arc<Mutex<Vec<StationTemperatures>>> =
+        Arc::new(Mutex::new(Vec::with_capacity(THREAD_COUNT)));
 
-    for stage_index in 0..stage_count {
-        let mut file_reader_threads = Vec::new();
+    (0..STAGE_COUNT).for_each(|stage_index| {
+        let mut file_reader_threads = Vec::with_capacity(THREAD_COUNT);
+
```

Comment:
`Vec::with_capacity()` definitely optimizes the vector expansion cost in runtime. 👌